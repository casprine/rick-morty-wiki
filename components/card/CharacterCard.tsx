import { FunctionComponent } from 'react';
import { styled } from '@/stitches';
import { Character } from '@/types';
import Image from 'next/image';

interface CharacterCardProps extends Character {}

export const CharacterCard: FunctionComponent<CharacterCardProps> = ({ location, image, name, status, species }) => {
  return (
    <Container>
      <div className="image-container">
        <Image src={image} alt={`${name}-image`} layout="fill" />
      </div>

      <div className="character-details">
        <div className="header">
          <div className="flex left">
            <h4 className="name">{name}</h4>
            <p className="species">{species}</p>
          </div>
          <Status status={status}>
            <p>{status}</p>
          </Status>
        </div>

        <p className="last-location">
          <span>Last seen at</span> <span className="location">{location.name}</span>
        </p>
      </div>
    </Container>
  );
};

const Container = styled('div', {
  padding: 10,
  border: '1px solid $borderColor',
  borderRadius: 4,
  cursor: 'pointer',

  '.image-container': {
    width: '100%',
    position: 'relative',
    height: '304px',
    borderRadius: 3,
    overflow: 'hidden',
  },

  '.character-details': {
    paddingTop: 10,

    '.header': {
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginTop: 10,
    },

    '.left': {
      flexDirection: 'column',
    },

    '.species': {
      fontSize: 15,
      color: '$gray600',
    },

    '.last-location': {
      margin: '10px 0',

      span: {
        fontSize: 16,
      },

      '.location': {
        color: '$black',
        fontWeight: 600,
      },
    },
  },
});

const Status = styled('div', {
  padding: '2px 10px',
  borderRadius: 8,

  variants: {
    status: {
      alive: {
        backgroundColor: '$greenLight',
        p: {
          color: '$green',
        },
      },

      dead: {
        backgroundColor: '$redLight',
        p: {
          color: '$red',
        },
      },

      unknown: {
        backgroundColor: '$grayLight',
        p: {
          color: '$gray',
        },
      },
    },
  },

  p: {
    fontSize: 12,
    fontWeight: 600,
    textTransform: 'uppercase',
  },
});
