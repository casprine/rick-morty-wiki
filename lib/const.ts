import { Character } from '@/types';

export const BASE_URL = `https://rickandmortyapi.com/api`;
export const BASE_CHARACTER_URL = `https://rickandmortyapi.com/api/character`;

export const testCharacter: Character = {
  name: 'Morty Smith',
  status: 'alive',
  species: 'Human',
  location: {
    name: 'Citadel of Ricks',
  },
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
};
