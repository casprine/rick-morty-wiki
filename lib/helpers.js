"use strict";
exports.__esModule = true;
exports.capitalize = void 0;
var capitalize = function (string) { return string.charAt(0).toUpperCase() + string.slice(1); };
exports.capitalize = capitalize;
