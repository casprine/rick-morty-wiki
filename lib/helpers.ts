import { Character, Episode, StatusType } from '@/types';

export const capitalize = (string: string) => string.charAt(0).toUpperCase() + string.slice(1);

export const isEmpty = (value: Array<any>) => value.length === 0;

export const transformCharacters = (data: Character[]) => {
  return data.map(({ name, status, species, location, gender, id, image }: Character): Character => {
    return {
      name,
      status: status.toLocaleLowerCase() as StatusType,
      species,
      location,
      gender,
      id,
      image,
    };
  });
};

export const transformCharacterEpisodes = (data: any[]): Array<Episode> => {
  return data.map(
    ({ id, name, episode, air_date }: { id: number; name: string; episode: string; air_date: string }) => {
      return {
        name,
        episode,
        airDate: air_date,
        id,
      };
    },
  );
};
