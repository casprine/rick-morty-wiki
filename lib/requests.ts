import { Character, CharacterResponse, Episode, GetCharacterParams } from '@/types';
import { BASE_URL, BASE_CHARACTER_URL } from './const';

// utils
const fetcher = (url: string) => fetch(url).then((r) => r.json());
const getEpisodeIds = (urls: string[] = []): number[] => {
  const ids = urls.map((url) => Number(url.split('/episode/')[1]));
  return ids;
};

/**
 * @param {string} id
 * @returns Character
 */
export const getSingleCharacter = async (id: string): Promise<Character> => {
  return await fetcher(`${BASE_URL}/character/${id}`);
};

/**
 * @param {GetCharacterParams} params
 * @returns Array of CharacterResponse
 */
export const getCharacters = async ({
  status = '',
  gender = '',
  pageNumber = 1,
  species = '',
}: GetCharacterParams): Promise<CharacterResponse> => {
  console.log({
    status,
    gender,
    pageNumber,
    species,
  });
  const url = `${BASE_CHARACTER_URL}/?page=${pageNumber}&status=${status}&gender=${gender}&species=${species}`;
  return await fetcher(url);
};

/**
 * @param {Array<String>} episode
 * @returns Array of Episodes | Array<Episode>
 */

export const getEpisodes = async (episodes: Array<string>): Promise<Array<Episode>> => {
  const ids = getEpisodeIds(episodes);
  const url = `${BASE_URL}/episode/${ids}`;
  const response = await fetcher(url);

  if (!Array.isArray(response)) return [response];
  return [...response];
};
