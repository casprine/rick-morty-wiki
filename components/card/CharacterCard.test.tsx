import { CharacterCard } from './CharacterCard';
import { render } from '@testing-library/react';
import { testCharacter } from 'lib/const';

test('renders a character card', () => {
  const component = render(<CharacterCard {...testCharacter} />);

  expect(component).toMatchSnapshot();
});
