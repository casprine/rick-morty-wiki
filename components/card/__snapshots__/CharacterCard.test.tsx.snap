// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a character card 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="c-ibSkSK"
      >
        <div
          class="image-container"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="Morty Smith-image"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="character-details"
        >
          <div
            class="header"
          >
            <div
              class="flex left"
            >
              <h4
                class="name"
              >
                Morty Smith
              </h4>
              <p
                class="species"
              >
                Human
              </p>
            </div>
            <div
              class="c-lmEuKC c-lmEuKC-bmSCHF-status-alive"
            >
              <p>
                alive
              </p>
            </div>
          </div>
          <p
            class="last-location"
          >
            <span>
              Last seen at
            </span>
             
            <span
              class="location"
            >
              Citadel of Ricks
            </span>
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="c-ibSkSK"
    >
      <div
        class="image-container"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Morty Smith-image"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="character-details"
      >
        <div
          class="header"
        >
          <div
            class="flex left"
          >
            <h4
              class="name"
            >
              Morty Smith
            </h4>
            <p
              class="species"
            >
              Human
            </p>
          </div>
          <div
            class="c-lmEuKC c-lmEuKC-bmSCHF-status-alive"
          >
            <p>
              alive
            </p>
          </div>
        </div>
        <p
          class="last-location"
        >
          <span>
            Last seen at
          </span>
           
          <span
            class="location"
          >
            Citadel of Ricks
          </span>
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
